frame/society/src/lib.rs

```diff
@@ -0,0 +1,1035 @@
+		pub Pot get(fn pot) config(): BalanceOf<T, I>;
+
+		/// The most primary from the most recently approved members.
+		pub Head get(head) build(|config: &GenesisConfig<T, I>| config.members.first().cloned()):
+			Option<T::AccountId>;
+
+		/// The current set of members, ordered.
+		pub Members get(fn members) build(|config: &GenesisConfig<T, I>| {
+			let mut m = config.members.clone();
+			m.sort();
+			m
+		}): Vec<T::AccountId>;
+
+		/// The set of suspended members.
+		pub SuspendedMembers get(fn suspended_member): map T::AccountId => Option<()>;
+
+		/// The current bids.
+		Bids: Vec<(BalanceOf<T, I>, T::AccountId, BidKind<T::AccountId, BalanceOf<T, I>>)>;
+
+		/// Members currently vouching, stored ordered
+		Vouching: Vec<T::AccountId>;
+
+		/// Pending payouts; ordered by block number, with the amount that should be paid out.
+		Payouts: map T::AccountId => Vec<(T::BlockNumber, BalanceOf<T, I>)>;
+
+		/// The ongoing number of losing votes cast by the member.
+		Strikes: map T::AccountId => StrikeCount;
+
+		/// Double map from Candidate -> Voter -> (Maybe) Vote.
+		Votes: double_map
+			hasher(twox_64_concat) T::AccountId,
+			twox_64_concat(T::AccountId)
+		=> Option<Vote>;
+
+		/// The defending member currently being challenged.
+		Defender get(fn defender): Option<T::AccountId>;
+		
+		/// Votes for the defender.
+		DefenderVotes: map hasher(twox_64_concat) T::AccountId => Option<Vote>;
+	}
+	add_extra_genesis {
+		config(members): Vec<T::AccountId>;
+	}
+}
+
+// The module's dispatchable functions.
+decl_module! {
+	/// The module declaration.
+	pub struct Module<T: Trait<I>, I: Instance=DefaultInstance> for enum Call where origin: T::Origin {
+		/// The minimum amount of a deposit required for a bid to be made.
+		const CandidateDeposit: BalanceOf<T, I> = T::CandidateDeposit::get();
+
+		/// The proportion of the unpaid reward that gets deducted in the case that either a skeptic
+		/// doesn't vote or someone votes in the wrong way.
+		const WrongSideDeduction: BalanceOf<T, I> = T::WrongSideDeduction::get();
+
+		/// The number of times a member may vote the wrong way (or not at all, when they are a skeptic)
+		/// before they become suspended.
+		const MaxStrikes: u32 = T::MaxStrikes::get();
+
+		/// The amount of incentive paid within each period. Doesn't include VoterTip.
+		const PeriodSpend: BalanceOf<T, I> = T::PeriodSpend::get();
+
+		/// The number of blocks between candidate/membership rotation periods.
+		const RotationPeriod: T::BlockNumber = T::RotationPeriod::get();
+
+		/// The number of blocks between membership challenges.
+		const ChallengePeriod: T::BlockNumber = T::ChallengePeriod::get();
+
+		// Used for handling module events.
+		fn deposit_event() = default;
+
+		/// Make a bid for entry.
+		///
+		/// Alters Bids (O(N) decode+encode, O(logN) search, <=2*O(N) write).
+		pub fn bid(origin, value: BalanceOf<T, I>) -> Result {
```

Comment:
I am worried that we do not check that a user may already be an existing bid. This would cause them to re-reserve some funds.

These reserved funds would not have any process to then be unreserved.

However, checking this might be expensive since we need to search the bids for an existing bid.